SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp

```diff
@@ -186,35 +167,141 @@ void ActorSystem::Advance()
+	{
+		return;
+	}
+
+	for (const EntityDelta& Delta : SubViewUpdate.EntityDeltas)
+	{
+		if (Delta.Type == EntityDelta::REMOVE || Delta.Type == EntityDelta::TEMPORARILY_REMOVED)
+		{
+			const Worker_EntityId EntityId = Delta.EntityId;
+			if (PresentEntities.Contains(EntityId))
+			{
+				const Worker_ComponentSetId AuthorityComponentSet = SubViewUpdate.SubViewType == ENetRole::ROLE_Authority
+																		? SpatialConstants::SERVER_AUTH_COMPONENT_SET_ID
+																		: SpatialConstants::CLIENT_AUTH_COMPONENT_SET_ID;
+
+				AuthorityLost(EntityId, AuthorityComponentSet);
+			}
+		}
+	}
+}
+
+ActorSystem::ActorSystem(const FSubView& InActorSubView, const FSubView& InAuthoritySubView, const FSubView& InOwnershipSubView,
+						 const FSubView& InSimulatedSubView, const FSubView& InTombstoneSubView, USpatialNetDriver* InNetDriver,
+						 SpatialEventTracer* InEventTracer)
+	: ActorSubView(&InActorSubView)
+	, AuthoritySubView(&InAuthoritySubView)
+	, OwnershipSubView(&InOwnershipSubView)
+	, SimulatedSubView(&InSimulatedSubView)
+	, TombstoneSubView(&InTombstoneSubView)
+	, NetDriver(InNetDriver)
+	, EventTracer(InEventTracer)
+	, ClaimPartitionHandler(*InNetDriver->Connection)
+{
+}
+
+#if DO_CHECK
+static void ValidateNoSubviewIntersections(const FSubView& Lhs, const FSubView& Rhs, const FString& SubviewDescription)
+{
+	for (const Worker_EntityId Overlapping : Lhs.GetCompleteEntities().Intersect(Rhs.GetCompleteEntities()))
+	{
+		UE_LOG(LogActorSystem, Warning, TEXT("Entity %lld is doubly complete on %s"), Overlapping, *SubviewDescription);
+	}
+}
+#endif // DO_CHECK
+
+void ActorSystem::Advance()
+{
+	for (const EntityDelta& Delta : ActorSubView->GetViewDelta().EntityDeltas)
+	{
+		if (Delta.Type == EntityDelta::REMOVE)
+		{
 			EntityRemoved(Delta.EntityId);
-			ActorDataStore.Remove(Delta.EntityId);
-			PopulateDataStore(Delta.EntityId);
-			EntityAdded(Delta.EntityId);
-			break;
-		default:
-			break;
+
+			const int32 EntitiesRemoved = PresentEntities.Remove(Delta.EntityId);
+
+			check(EntitiesRemoved != 0);
 		}
 	}
 
+	struct FEntitySubView
+	{
+		const FSubView* SubView;
+		ENetRole Type;
+
+		operator FEntitySubViewUpdate() const { return { SubView->GetViewDelta().EntityDeltas, Type }; }
+	};
+
+#if DO_CHECK
+	{
+		ValidateNoSubviewIntersections(*AuthoritySubView, *OwnershipSubView, TEXT("Authority and Ownership"));
+		ValidateNoSubviewIntersections(*AuthoritySubView, *SimulatedSubView, TEXT("Authority and Simulated"));
+		ValidateNoSubviewIntersections(*SimulatedSubView, *OwnershipSubView, TEXT("Simulated and Ownership"));
+	}
+#endif // DO_CHECK
+
+	const FEntitySubView SubViews[]{
+		{ AuthoritySubView, ENetRole::ROLE_Authority },
+		{ OwnershipSubView, ENetRole::ROLE_AutonomousProxy },
+		{ SimulatedSubView, ENetRole::ROLE_SimulatedProxy },
+	};
+
+	for (const FEntitySubView& SubView : SubViews)
+	{
+		ProcessRemoves(SubView);
+	}
+
+	for (const FEntitySubView& SubView : SubViews)
+	{
+		ProcessUpdates(SubView);
```

Comment:
These functions make it fairly implicit what's happening inside them, and you're depending on the function to handle whatever set of subviews happen to be in the list (which is stored outside the function). What's the reason for doing it this way? To avoid duplicating code? 

Personally I would prefer processing each subview explicitly if it's not too unwieldy, to minimise the cognitive overhead of changing these functions/the set of subviews in future.